File: iris-ld/iris-ld-jsonld/src/main/java/city/smartb/iris/jsonld/reader/JsonField.java
package city.smartb.iris.jsonld.reader;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

public abstract class JsonField {

    protected final Object obj;

    public JsonField(Object obj) {
        this.obj = obj;
    }

    public JsonField(Map<String, Object> json, String key) {
        this.obj = json.get(key);
    }

    public abstract String asString();

    public abstract Integer asInteger();

    public abstract Map<String, Object> asMap();

    public abstract List<Map<String, Object>> getMaps();

    public abstract <T> T asObject(Class<T> clazz);

    public abstract <T> List<T> asListObjects(Class<T> clazz);

    public abstract LocalDate asLocalDate();

    public abstract LocalDateTime asLocalDateTime();
}

File: iris-ld/iris-ld-jsonld/src/main/kotlin/city/smartb/iris/ld/jsonld/reader/JsonField.kt
package city.smartb.iris.ld.jsonld.reader

import java.time.LocalDate
import java.time.LocalDateTime

abstract class JsonField {
    protected val obj: Any?

    constructor(obj: Any) {
        this.obj = obj
    }

    constructor(json: Map<String, Any>, key: String?) {
        obj = json[key]
    }

    abstract fun asString(): String?
    abstract fun asInteger(): Int
    abstract fun asMap(): Map<String, Any>
    abstract val maps: List<Map<String, Any>>
    abstract fun <T> asObject(clazz: Class<T>?): T
    abstract fun <T> asListObjects(clazz: Class<T>?): List<T>
    abstract fun asLocalDate(): LocalDate
    abstract fun asLocalDateTime(): LocalDateTime
}
